### PR TITLE
runtest: add cProfile support

### PR DESCRIPTION
**Pull request by jd**
This patch allows to profile unit tests using cProfile. If the
environment variable TESTTOOLS_CPROFILE is set, a `cprof` file is
generated with the full name of the test, containing the profiling
information.

[Review on Reviewable](https://reviewable.io/reviews/testing-cabal/testtools/163)

### CODE REVIEW

**Comment by jml** [comment]
@jd Thanks for the patch. 

What were your reasons for implementing it in `RunTest`? 

I ask because I think it should be something separate. Perhaps either as something that wraps an existing `RunTest` (although I've just tried this recently, and it's a lot harder & fiddlier than it should be), or as a `Fixture`.

File: testtools/runtest.py [hunk 1]
```diff
@@ -76,9 +83,17 @@ def run(self, result=None):
...
+            self._pr.enable()
         try:
             return self._run_one(actual_result)
         finally:
+            if self._pr:
+                self._pr.disable()
+                self._pr.dump_stats("%s.%s.%s.cprof"
+                                    % (self.case.__class__.__module__,
+                                       self.case.__class__.__name__,
+                                       self.case._get_test_method().__name__))
```
Comment (jml) [inline testtools/runtest.py]: This should be attached as a detail. AIUI, right now it just prints to stdout.

**Comment by jd** [comment]
Hi @jml! This is a very naive implementation, so any suggestion welcome. I'd love to have that automagically available to any project that uses testtools, so that's why I implemented it this way.

I don't think using a `Fixture` would allow that?

**Comment by rbtcollins** [comment]
@jd thanks for the patch. I think using an environment variable to activate this is a bi of a rough UI. I'd rather see testtools.run grow options to enable / disable profiling.

jono's comments about code structure don't really alter wether this is builtin or not, but will help keep concerns that should be separate separate.

File: testtools/runtest.py (hunk 1 above)
Comment (jd) [inline testtools/runtest.py]: @jml no, it prints it out to a file. But attaching to a result sounds like a good idea.

**Comment by jml** [comment]
@jd Thanks.

@rbtcollins is right in that 

Putting together what @rbtcollins and I suggested:
* Please move the logic to a `Fixture`
* Have that fixture return the data as a detail
*

**Comment by jml** [comment]
I keep hitting the wrong button!

**Comment by rbtcollins** [comment]
I'm not clear; this seems like we've said 'you should do this externally'? If so, may we close the PR?

**Comment by jml** [comment]
I guess I'd be OK with merging in a fixture that does what this does, but probably it would be better for that to be implemented as part of fixtures itself, or its own library. So, yeah, close for now.

**Pull request closed without merging**